### PR TITLE
Issue 91 average diffusivity

### LINKED ISSUE #91: Discretising "diffusivity times gradient" in Finite Volumes
Flux is of the form `N = -D(c) * grad(c)`

With the current implementation, after discretisation, `grad(c)` will evaluate to a vector of size (n-1,) (at cell edges) while `D(c)` (an arbitrary function of `c`, e.g. `c**2`) will evaluate to a vector of size (n,) (at cell centres), where (n,) is the size of c. So trying to multiply these two will give an error.

The Finite Volume discretisation needs to figure out that we have to take the average of `D(c)` to give `D_edges` which is a vector of size (n-1,) at cell edges. This could probably be done at the `BinaryOperator` stage of `disc.process_symbol`, i.e. if `right` is a `Gradient` object then take the average of `left`.

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Fixes #91 
Have re-opened #91 to fix the case where there is both a `divergence` and a `gradient`:
```
Vector * Gradient(Variable) # should do node_to_edge
Vector * Divergence(Gradient(Variable)) # should not do node_to_edge
```
In doing so, realised there was an issue in `BaseDiscretisation.process_symbol` where, in the final `else`, detaching then reattaching `symbol.parent` switches around the order of `parent.children`:
```
# before disc:
symbol = Scalar * Vector 
symbol.children -> (Scalar, Vector)
# after disc 
disc.process_symbol(symbol)
symbol.children -> (Vector, Scalar)
```
This is a problem for recognising `symbol.id` in the boundary conditions

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pybamm/discretisations/base_discretisation.py [hunk 1]
```diff
@@ -200,15 +199,22 @@ def process_symbol(self, symbol, y_slices=None, boundary_conditions={}):
 
             return new_symbol
 
+        elif isinstance(symbol, pybamm.Scalar):
+            return pybamm.Scalar(symbol.value, domain=symbol.domain)
+
+        elif isinstance(symbol, pybamm.Array):
+            return symbol.__class__(symbol.entries, domain=symbol.domain)
```
Comment (martinjrobins) [inline pybamm/discretisations/base_discretisation.py]:
```suggestion
            return symbol.Array(symbol.entries, domain=symbol.domain)
```

Comment (martinjrobins) [reply pybamm/discretisations/base_discretisation.py]: not sure what is better, but should be consistent with Scalar elif above

Comment (tinosulzer) [reply pybamm/discretisations/base_discretisation.py]: `Array` can be either `Vector` or `Matrix`, would be better to return the same type I think

Comment (martinjrobins) [reply pybamm/discretisations/base_discretisation.py]: ok, forgot about that. Ignore that suggestion then!
